<?xml version="1.0"?>

<robot
  name="quadrotor"
  xmlns:xacro="http://www.ros.org/wiki/xacro"
>

    <xacro:arg name="rate" default="10"/>
    <xacro:arg name="min_range" default="0.01"/>
    <xacro:arg name="max_range" default="10.0"/>

    <xacro:property name="M_PI" value="3.1415926535897931" />

    <!-- Included URDF Files -->
    <xacro:include filename="$(find quadrotor_tunnel_nav)/urdf/quadrotor_simple.urdf.xacro" />

    <!-- Instantiate quadrotor_base_macro once (has no parameters atm) -->
    <!-- page: http://wiki.ros.org/hector_gazebo_plugins#GazeboRosSonar -->
    <xacro:quadrotor_base_macro />

    <!-- ray sensors -->
    <xacro:include filename="$(find hector_sensors_description)/urdf/sonar_sensor.urdf.xacro" />
    <xacro:property name="RATE" value="$(arg rate)" />
    <xacro:property name="MIN" value="$(arg min_range)" />
    <xacro:property name="MAX" value="$(arg max_range)" />
    <xacro:property name="DEG" value="3" />
    <xacro:property name="STEP_X" value="0.05" />
    <xacro:property name="STEP_Y" value="0.05" />
    <xacro:property name="STEP_Z1" value="0.05" />
    <xacro:property name="STEP_Z2" value="0.05" />
    <!-- front -->
    <xacro:sonar_sensor name="ray_front" parent="base_link" ros_topic="range_front" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${STEP_X} 0.0 ${STEP_Z1}" rpy="0 0 0" />
    </xacro:sonar_sensor>
    <!-- right front -->
    <xacro:sonar_sensor name="ray_rfront" parent="base_link" ros_topic="range_rfront" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${STEP_X} ${-STEP_Y} ${STEP_Z1}" rpy="0 0 ${-45*M_PI/180}" />
    </xacro:sonar_sensor>
    <!-- left front -->
    <xacro:sonar_sensor name="ray_lfront" parent="base_link" ros_topic="range_lfront" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${STEP_X} ${STEP_Y} ${STEP_Z1}" rpy="0 0 ${45*M_PI/180}" />
    </xacro:sonar_sensor>
    <!-- right -->
    <xacro:sonar_sensor name="ray_right" parent="base_link" ros_topic="range_right" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="0 ${-STEP_Y} ${STEP_Z1}" rpy="0 0 ${-90*M_PI/180}" />
    </xacro:sonar_sensor>
    <!-- left -->
    <xacro:sonar_sensor name="ray_left" parent="base_link" ros_topic="range_left" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="0 ${STEP_Y} ${STEP_Z1}" rpy="0 0 ${90*M_PI/180}" />
    </xacro:sonar_sensor>
    <!-- right rear -->
    <xacro:sonar_sensor name="ray_rrear" parent="base_link" ros_topic="range_rrear" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${-STEP_X} ${-STEP_Y} ${STEP_Z1}" rpy="0 0 ${-135*M_PI/180}" />
    </xacro:sonar_sensor>
    <!-- left rear -->
    <xacro:sonar_sensor name="ray_lrear" parent="base_link" ros_topic="range_lrear" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${-STEP_X} ${STEP_Y} ${STEP_Z1}" rpy="0 0 ${135*M_PI/180}" />
    </xacro:sonar_sensor>
    <!-- rear -->
    <xacro:sonar_sensor name="ray_rear" parent="base_link" ros_topic="range_rear" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${-STEP_X} 0 ${STEP_Z1}" rpy="0 0 ${-180*M_PI/180}" />
    </xacro:sonar_sensor>

    <!-- upward -->
    <xacro:sonar_sensor name="ray_up" parent="base_link" ros_topic="range_up" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="0 0 ${STEP_Z1}" rpy="0 ${-90*M_PI/180} 0" />
    </xacro:sonar_sensor>
    <!-- up front -->
    <xacro:sonar_sensor name="ray_ufront" parent="base_link" ros_topic="range_ufront" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${STEP_X/3} 0 ${STEP_Z1}" rpy="0 ${-45*M_PI/180} 0" />
    </xacro:sonar_sensor>
    <!-- up rear -->
    <xacro:sonar_sensor name="ray_urear" parent="base_link" ros_topic="range_urear" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${-STEP_X/3} 0 ${STEP_Z1}" rpy="0 ${-45*M_PI/180} ${180*M_PI/180}" />
    </xacro:sonar_sensor>

    <!-- downward -->
    <xacro:sonar_sensor name="ray_down" parent="base_link" ros_topic="range_down" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="0 0 ${-STEP_Z2}" rpy="0 ${90*M_PI/180} 0" />
    </xacro:sonar_sensor>
    <!-- down front -->
    <xacro:sonar_sensor name="ray_dfront" parent="base_link" ros_topic="range_dfront" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${STEP_X/3} 0 ${-STEP_Z2}" rpy="0 ${45*M_PI/180} 0" />
    </xacro:sonar_sensor>
    <!-- down rear -->
    <xacro:sonar_sensor name="ray_drear" parent="base_link" ros_topic="range_drear" update_rate="${RATE}" min_range="${MIN}" max_range="${MAX}" field_of_view="${DEG*M_PI/180}" ray_count="1">
      <origin xyz="${-STEP_X/3} 0 ${-STEP_Z2}" rpy="0 ${45*M_PI/180} ${180*M_PI/180}" />
    </xacro:sonar_sensor>

    <!-- Hokuyo UTM-30LX mounted upside down below the quadrotor body
    <xacro:include filename="$(find hector_sensors_description)/urdf/hokuyo_utm30lx.urdf.xacro" />
    <xacro:hokuyo_utm30lx name="laser_h" parent="base_link" ros_topic="scan" update_rate="40" ray_count="1081" min_angle="-135" max_angle="135">
      <origin xyz="0.0 0.0 0.097" rpy="0 0 0"/>
    </xacro:hokuyo_utm30lx>
    -->

    <!-- Forward facing camera
    <xacro:include filename="$(find hector_sensors_description)/urdf/generic_camera.urdf.xacro" />
    <xacro:generic_camera name="front_cam" parent="base_link" ros_topic="camera/image" cam_info_topic="camera/camera_info" update_rate="10" res_x="320" res_y="240" image_format="R8G8B8" hfov="90">
      <origin xyz="0.05 0.0 -0.06" rpy="0 0 0"/>
    </xacro:generic_camera> -->

</robot>
